### PR TITLE
NVQC API versioning updates and advance REST API to v1.1

### PR DESCRIPTION
**Pull request by bmhowe23**
This PR updates the client and server API handling code to better handle advancing the capabilities of the server without having to update the client. The basic use case of this will be adding new features to the server while allowing users to keep using old client versions (if desired). As long as the server updates are done in a non-breaking way, we would simply advance the REST API version on the server from 1.0 to 1.1, 1.1 to 1.2, etc. If there is a breaking change on the server where old clients could not talk to the new server, we would advance the REST API version to 2.0.

For changes like the recent VQE and optimizer updates, the client can retrieve the server information in order to figure out whether the "new" way is available. If the "new" way is unavailable, then it will simply fall back on the slower (but still functional) old way.

Note: the `remote-mqpu` target behaves as if the client and server are essentially at the same version. That is - the client assumes that the server has all the optional capabilities that the client knows about. It behaved like that before, and this PR does not change that.

The 1st commit of this PR actually fixes a bug where the NVQC VQE updates from #1901 were not working correctly for the actual NVQC target. The 2nd and 3rd commits reduce some duplicated code in PyRemoteSimulatorQPU.cpp in order to make it less like that the oversight in #1901 happens again. I can pull all 3 of those commits into a separate PR if desired.

| REST API Version | CUDA-Q Client Version | CUDA-Q Server Version |
|----|----|-----|
| 1 (i.e. 1.0) | 0.7.0, 0.7.1 | 0.7.0 |
| 1.1 | 0.8.0 | 0.8.0 |

### CODE REVIEW

File: python/runtime/utils/PyRemoteSimulatorQPU.cpp [hunk 1]
```diff
@@ -14,6 +14,66 @@ using namespace mlir;
 
 namespace {
 
+// This is a helper function to help reduce duplicated code across
+// PyRemoteSimulatorQPU and PyNvcfSimulatorQPU.
+static void _launchVQE(cudaq::ExecutionContext *executionContextPtr,
```
[the remark this answers is not in the document]
Comment (bmhowe23) [reply python/runtime/utils/PyRemoteSimulatorQPU.cpp:19]: > Curious why underscore before the function name in cpp?

Just to signify that it is an internal function and that it is related to the `launchVQE` functions.

File: runtime/common/BaseRestRemoteClient.h [hunk 2]
```diff
@@ -436,9 +442,12 @@ class BaseNvcfRuntimeClient : public cudaq::BaseRemoteRestRuntimeClient {
   // Information about function deployment from environment variable info.
   struct FunctionEnvironments {
     // These configs should be positive numbers.
-    int version{-1};
+    int majorVersion{-1};
+    int minorVersion{-1};
```
[the remark this answers is not in the document]
Comment (bmhowe23) [reply runtime/common/BaseRestRemoteClient.h:446]: > Should we add a patchVersion{-1} here as well?

The major and minor versions correspond to the API, not a given software release. If we update the API, I suspect that would be a bump to the minor version, not a patch version.

File: runtime/common/BaseRestRemoteClient.h [hunk 3]
```diff
@@ -757,6 +839,43 @@ class BaseNvcfRuntimeClient : public cudaq::BaseRemoteRestRuntimeClient {
...
+  // The NVCF version of this function needs to dynamically determine the remote
+  // capabilities based on the servers currently deployed.
+  virtual RemoteCapabilities getRemoteCapabilities() const override {
+    // If an environment variable override is activated, then enable all remote
+    // capabilities.
+    if (auto *envVal = std::getenv("CUDAQ_CLIENT_REMOTE_CAPABILITY_OVERRIDE")) {
+      std::string tmp(envVal);
+      std::transform(tmp.begin(), tmp.end(), tmp.begin(),
+                     [](unsigned char c) { return std::tolower(c); });
+      if (tmp == "1" || tmp == "on" || tmp == "true" || tmp == "yes")
```
[the remark this answers is not in the document]
Comment (bmhowe23) [reply runtime/common/BaseRestRemoteClient.h:852]: > I think we should allow only one type of value to be set ("true").

This is following the precedence other places in our code (Base RemoteRESTQPU.h and BaseRemoteSimulatorQPU.h).

File: runtime/common/BaseRestRemoteClient.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (bmhowe23) [reply runtime/common/BaseRestRemoteClient.h:446]: That would bump the software version but not the API version, no?

File: runtime/common/BaseRestRemoteClient.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (bmhowe23) [reply runtime/common/BaseRestRemoteClient.h:852]: I think it is beneficial to allow users the flexibility to specify flags in multiple ways, similar to how both python and C++ don't limit `if` checks to purely boolean values. (I.e. they allow integers as well.) Python also has a similar function that supports decoding multiple strings into boolean values: https://docs.python.org/3.11/distutils/apiref.html#distutils.util.strtobool.

File: runtime/common/BaseRestRemoteClient.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (bmhowe23) [reply runtime/common/BaseRestRemoteClient.h:446]: If the bug fix doesn't directly affect the API, then the major and minor API version should remain unchanged. If, on the other hand, the bug fix is _related_ to the API, I anticipate that we would bump the minor version of the API. So, unless I'm missing something, I don't think that we would need a patch version for any P0 bug fixes.

**Review by 1tnguyen — APPROVED** [review]
LGTM 👍

**Comment by bmhowe23** [comment]
Note: if #1958 is merged prior to this PR, then remove the vqe_h2.cpp exception that was added to the integration tests because this PR fixes that issue.

**Review by bettinaheim — APPROVED** [review]

**Pull request merged**